File: economy-plugin/src/main/java/me/derechtepilz/economy/commands/EconomyCommand.java
package me.derechtepilz.economy.commands;

import dev.jorel.commandapi.CommandTree;
import dev.jorel.commandapi.arguments.*;
import me.derechtepilz.economy.Main;
import me.derechtepilz.economy.permissionmanagement.Permission;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.permissions.ServerOperator;

import static dev.jorel.commandapi.arguments.LiteralArgument.of;

public class EconomyCommand {

    private final EconomyCommandExecution commandExecution;

    public EconomyCommand(Main main) {
        commandExecution = new EconomyCommandExecution(main);
    }

    public void register() {
        new CommandTree("economy")
            .then(of("auction")
                .then(of("buy")
                    .executesPlayer(commandExecution::buyAuction)
                    .then(new ItemStackArgument("filter")
                        .executesPlayer(commandExecution::buyAuctionWithFilter)
                    )
                )
                .then(of("create")
                    .then(new ItemStackArgument("item")
                        .then(new IntegerArgument("amount", 1)
                            .then(new DoubleArgument("price", 1)
                                .then(new IntegerArgument("hours", 0)
                                    .then(new IntegerArgument("minutes", 0)
                                        .then(new IntegerArgument("seconds", 0)
                                            .executesPlayer(commandExecution::createAuction)
                                        )
                                    )
                                )
                            )
                        )
                    )
                )
                .then(of("cancel")
                    .executesPlayer(commandExecution::cancelAuction)
                )
                .then(of("claim")
                    .executesPlayer(commandExecution::claimAuction)
                )
                .then(of("pause")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.PAUSE_RESUME_AUCTIONS) || player.isOp();
                        }
                        return false;
                    })
                    .executesPlayer(commandExecution::pauseAuction)
                )
                .then(of("resume")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.PAUSE_RESUME_AUCTIONS) || player.isOp();
                        }
                        return false;
                    })
                    .executesPlayer(commandExecution::resumeAuction)
                )
            )
            .then(of("coins")
                .then(of("give")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.GIVE_COINS) || player.isOp();
                        }
                        return false;
                    })
                    .then(new PlayerArgument("target")
                        .then(new DoubleArgument("amount", 0)
                            .executesPlayer(commandExecution::giveCoins)
                        )
                    )
                )
                .then(of("take")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.TAKE_COINS) || player.isOp();
                        }
                        return false;
                    })
                    .then(new PlayerArgument("target")
                        .then(new DoubleArgument("amount", 0)
                            .executesPlayer(commandExecution::takeCoins)
                        )
                    )
                )
                .then(of("set")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.SET_COINS) || player.isOp();
                        }
                        return false;
                    })
                    .then(new PlayerArgument("target")
                        .then(new DoubleArgument("amount", 0)
                            .executesPlayer(commandExecution::setCoins)
                        )
                    )
                )
                .then(of("baltop")
                    .executesPlayer(commandExecution::balTop)
                )
            )
            .then(of("permission")
                .withRequirement(ServerOperator::isOp)
                .then(of("clear")
                    .then(new PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings(suggestionInfo -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)))
                        .executesPlayer(commandExecution::clearPermissions)
                    )
                )
                .then(of("single")
                    .then(of("set")
                        .then(new PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings(suggestionInfo -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)))
                            .then(new ListArgumentBuilder<String>("permissions", ",").allowDuplicates(false).withList(Permission.getPermissions()).withStringMapper().build()
                                .withRequirement(ServerOperator::isOp)
                                .executesPlayer(commandExecution::setSinglePermission)
                            )
                        )
                    )
                    .then(of("remove")
                        .then(new PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings(suggestionInfo -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)))
                            .then(new ListArgumentBuilder<String>("permissions", ",").allowDuplicates(false).withList(Permission.getPermissions()).withStringMapper().build()
                                .withRequirement(ServerOperator::isOp)
                                .executesPlayer(commandExecution::removeSinglePermission)
                            )
                        )
                    )
                    .then(of("get")
                        .then(new PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings(suggestionInfo -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)))
                            .then(new ListArgumentBuilder<String>("permissions", ",").allowDuplicates(false).withList(Permission.getPermissions()).withStringMapper().build()
                                .withRequirement(ServerOperator::isOp)
                                .executesPlayer(commandExecution::getSinglePermission)
                            )
                        )
                    )
                )
            )
            .then(of("friend")
                .then(of("add")
                    .then(new PlayerArgument("target").replaceSuggestions(ArgumentSuggestions.strings(suggestionInfo -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)))
                        .executesPlayer(commandExecution::addFriend)
                    )
                )
                .then(of("remove")
                    .then(new PlayerArgument("target").replaceSuggestions(ArgumentSuggestions.strings(suggestionInfo -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)))
                        .executesPlayer(commandExecution::removeFriend)
                    )
                )
                .then(of("accept")
                    .then(new PlayerArgument("target").replaceSuggestions(ArgumentSuggestions.strings(suggestionInfo -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)))
                        .executesPlayer(commandExecution::acceptFriend)
                    )
                )
                .then(of("deny")
                    .then(new PlayerArgument("target").replaceSuggestions(ArgumentSuggestions.strings(suggestionInfo -> Bukkit.getOnlinePlayers().stream().map(Player::getName).toArray(String[]::new)))
                        .executesPlayer(commandExecution::denyFriend)
                    )
                )
            )
            .then(of("config")
                .then(of("allowDirectDownloads")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.MODIFY_CONFIG) || player.isOp();
                        }
                        return false;
                    })
                    .then(new BooleanArgument("allowDirectDownloads")
                        .executesPlayer(commandExecution::allowDirectDownloads)
                    )
                )
                .then(of("startBalance")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.MODIFY_CONFIG) || player.isOp();
                        }
                        return false;
                    })
                    .then(new DoubleArgument("startBalance", 0.0)
                        .executesPlayer(commandExecution::startBalance)
                    )
                )
                .then(of("interestRate")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.MODIFY_CONFIG) || player.isOp();
                        }
                        return false;
                    })
                    .then(new DoubleArgument("interestRate", 0.0)
                        .executesPlayer(commandExecution::interestRate)
                    )
                )
                .then(of("minimumDaysForInterest")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.MODIFY_CONFIG) || player.isOp();
                        }
                        return false;
                    })
                    .then(new IntegerArgument("minimumDaysForInterest", 1)
                        .executesPlayer(commandExecution::minimumDaysForInterest)
                    )
                )
                .then(of("reset")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.RESET_CONFIG) || player.isOp();
                        }
                        return false;
                    })
                    .executesPlayer(commandExecution::resetConfig)
                )
                .then(of("reload")
                    .withRequirement(sender -> {
                        if (sender instanceof Player player) {
                            return Permission.hasPermission(player, Permission.RESET_CONFIG) || player.isOp();
                        }
                        return false;
                    })
                    .executesPlayer(commandExecution::reloadConfig)
                )
            )
            .register();
    }
}

File: economy-plugin/src/main/java/io/github/derechtepilz/economy/commands/EconomyCommand.kt
package io.github.derechtepilz.economy.commands

import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.arguments.ListArgumentBuilder
import dev.jorel.commandapi.arguments.LiteralArgument.of
import dev.jorel.commandapi.arguments.PlayerArgument
import dev.jorel.commandapi.arguments.StringArgument
import io.github.derechtepilz.economy.Main
import io.github.derechtepilz.economy.permissionmanagement.Permission
import io.github.derechtepilz.economy.permissionmanagement.PermissionGroup
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class EconomyCommand(main: Main) {
    private val commandExecution: EconomyCommandExecution = EconomyCommandExecution(main)

    fun register() {
        commandTree("economy") {
            literalArgument("auction") {
                literalArgument("buy") {
                    playerExecutor { player, args ->
                        commandExecution.buyAuction(player, args)
                    }
                    itemStackArgument("filter") {
                        playerExecutor { player, args ->
                            commandExecution.buyAuctionWithFilter(player, args)
                        }
                    }
                }
                literalArgument("create") {
                    itemStackArgument("item") {
                        integerArgument("amount", 1) {
                            doubleArgument("price", 1.0) {
                                integerArgument("hours", 0) {
                                    integerArgument("minutes", 0) {
                                        integerArgument("seconds", 0) {
                                            playerExecutor { player, args ->
                                                commandExecution.createAuction(player, args)
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
                literalArgument("cancel") {
                    playerExecutor { player, args ->
                        commandExecution.cancelAuction(player, args)
                    }
                }
                literalArgument("claim") {
                    playerExecutor { player, args ->
                        commandExecution.claimAuction(player, args)
                    }
                }
                requirement(of("pause"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.PAUSE_RESUME_AUCTIONS) || sender.isOp }) {
                    playerExecutor { player, args ->
                        commandExecution.pauseAuction(player, args)
                    }
                }
                requirement(of("resume"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.PAUSE_RESUME_AUCTIONS) || sender.isOp }) {
                    playerExecutor { player, args ->
                        commandExecution.resumeAuction(player, args)
                    }
                }
            }
            literalArgument("coins") {
                requirement(of("give"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.GIVE_COINS) || sender.isOp }) {
                    playerArgument("target") {
                        doubleArgument("amount", 0.0) {
                            playerExecutor { player, args ->
                                commandExecution.giveCoins(player, args)
                            }
                        }
                    }
                }
                requirement(of("take"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.TAKE_COINS) || sender.isOp }) {
                    playerArgument("target") {
                        doubleArgument("amount", 0.0) {
                            playerExecutor { player, args ->
                                commandExecution.takeCoins(player, args)
                            }
                        }
                    }
                }
                requirement(of("set"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.SET_COINS) || sender.isOp }) {
                    playerArgument("target") {
                        doubleArgument("amount", 0.0) {
                            playerExecutor { player, args ->
                                commandExecution.setCoins(player, args)
                            }
                        }
                    }
                }
                literalArgument("baltop") {
                    playerExecutor { player, args ->
                        commandExecution.balTop(player, args)
                    }
                }
            }
            requirement(of("permission"), { obj: CommandSender -> obj.isOp }) {
                literalArgument("clear") {
                    argument(PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings { Bukkit.getOnlinePlayers().stream().map { obj: Player -> obj.name }.toList().toTypedArray() })) {
                        playerExecutor { player, args ->
                            commandExecution.clearPermissions(player, args)
                        }
                    }
                }
                literalArgument("single") {
                    literalArgument("set") {
                        argument(PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings{ Bukkit.getOnlinePlayers().stream().map { obj: Player -> obj.name }.toList().toTypedArray() })) {
                            argument(ListArgumentBuilder<String>("permission", ",").allowDuplicates(false).withList(Permission.getPermissions()).withStringMapper().buildGreedy()) {
                                playerExecutor { player, args ->
                                    commandExecution.setSinglePermission(player, args)
                                }
                            }
                        }
                    }
                    literalArgument("remove") {
                        argument(PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings{ Bukkit.getOnlinePlayers().stream().map { obj: Player -> obj.name }.toList().toTypedArray() })) {
                            argument(ListArgumentBuilder<String>("permission", ",").allowDuplicates(false).withList(Permission.getPermissions()).withStringMapper().buildGreedy()) {
                                playerExecutor { player, args ->
                                    commandExecution.removeSinglePermission(player, args)
                                }
                            }
                        }
                    }
                }
                literalArgument("group") {
                    literalArgument("set") {
                        argument(PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings { Bukkit.getOnlinePlayers().stream().map { obj: Player -> obj.name }.toList().toTypedArray() })) {
                            argument(ListArgumentBuilder<String>("permissionGroup", ",").allowDuplicates(false).withList(PermissionGroup.getPermissionGroups()).withStringMapper().buildGreedy()) {
                                playerExecutor { player, args ->
                                    commandExecution.setPermissionGroup(player, args)
                                }
                            }
                        }
                    }
                    literalArgument("remove") {
                        argument(PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings { Bukkit.getOnlinePlayers().stream().map { obj: Player -> obj.name }.toList().toTypedArray() })) {
                            argument(ListArgumentBuilder<String>("permissionGroup", ",").allowDuplicates(false).withList(PermissionGroup.getPermissionGroups()).withStringMapper().buildGreedy()) {
                                playerExecutor { player, args ->
                                    commandExecution.removePermissionGroup(player, args)
                                }
                            }
                        }
                    }
                    literalArgument("register") {
                        stringArgument("name") {
                            argument(ListArgumentBuilder<String>("permissions", ",").allowDuplicates(false).withList(Permission.getPermissions()).withStringMapper().buildGreedy()) {
                                playerExecutor { player, args ->
                                    commandExecution.registerPermissionGroup(player, args)
                                }
                            }
                        }
                    }
                    literalArgument("delete") {
                        argument(StringArgument("permissionGroup").replaceSuggestions(ArgumentSuggestions.strings { PermissionGroup.getPermissionGroups().toTypedArray() })) {
                            playerExecutor { player, args ->
                                commandExecution.deletePermissionGroup(player, args)
                            }
                        }
                    }
                }
                literalArgument("get") {
                    argument(PlayerArgument("player").replaceSuggestions(ArgumentSuggestions.strings { Bukkit.getOnlinePlayers().stream().map { obj: Player -> obj.name }.toList().toTypedArray() })) {
                        playerExecutor { player, args ->
                            commandExecution.getSinglePermission(player, args)
                        }
                    }
                }
            }
            literalArgument("friend") {
                literalArgument("add") {
                    argument(PlayerArgument("target").replaceSuggestions(ArgumentSuggestions.strings {Bukkit.getOnlinePlayers().stream().map { player: Player -> player.name }.toList().toTypedArray()})) {
                        playerExecutor { player, args ->
                            commandExecution.addFriend(player, args)
                        }
                    }
                }
                literalArgument("remove") {
                    argument(PlayerArgument("target").replaceSuggestions(ArgumentSuggestions.strings {Bukkit.getOnlinePlayers().stream().map { player: Player -> player.name }.toList().toTypedArray()})) {
                        playerExecutor { player, args ->
                            commandExecution.removeFriend(player, args)
                        }
                    }
                }
                literalArgument("accept") {
                    argument(PlayerArgument("target").replaceSuggestions(ArgumentSuggestions.strings {Bukkit.getOnlinePlayers().stream().map { player: Player -> player.name }.toList().toTypedArray()})) {
                        playerExecutor { player, args ->
                            commandExecution.acceptFriend(player, args)
                        }
                    }
                }
                literalArgument("deny") {
                    argument(PlayerArgument("target").replaceSuggestions(ArgumentSuggestions.strings {Bukkit.getOnlinePlayers().stream().map { player: Player -> player.name }.toList().toTypedArray()})) {
                        playerExecutor { player, args ->
                            commandExecution.denyFriend(player, args)
                        }
                    }
                }
            }
            literalArgument("config") {
                requirement(of("allowDirectDownloads"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.MODIFY_CONFIG) || sender.isOp }) {
                    booleanArgument("allowDirectDownloads") {
                        playerExecutor { player, args ->
                            commandExecution.allowDirectDownloads(player, args)
                        }
                    }
                }
                requirement(of("startBalance"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.MODIFY_CONFIG) || sender.isOp }) {
                    doubleArgument("startBalance", 0.0) {
                        playerExecutor { player, args ->
                            commandExecution.startBalance(player, args)
                        }
                    }
                }
                requirement(of("interestRate"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.MODIFY_CONFIG) || sender.isOp }) {
                    doubleArgument("interestRate", 0.0) {
                        playerExecutor { player, args ->
                            commandExecution.interestRate(player, args)
                        }
                    }
                }
                requirement(of("minimumDaysForInterest"), { sender: CommandSender -> sender is Player && Permission.hasPermission(sender, Permission.MODIFY_CONFIG) || sender.isOp }) {
                    integerArgument("minimumDaysForInterest", 1) {
                        playerExecutor { player, args ->
                            commandExecution.minimumDaysForInterest(player, args)
                        }
                    }
                }
                requirement(of("reset"), { sender: CommandSender -> (sender is Player) && Permission.hasPermission(sender, Permission.RESET_CONFIG) || sender.isOp }) {
                    playerExecutor { player, args ->
                        commandExecution.resetConfig(player, args)
                    }
                }
                requirement(of("reload"), { sender: CommandSender -> (sender is Player) && Permission.hasPermission(sender, Permission.RESET_CONFIG) || sender.isOp }) {
                    playerExecutor { player, args ->
                        commandExecution.reloadConfig(player, args)
                    }
                }
            }
        }
    }
}